File: src/test/java/no/nav/familie/integrasjoner/personopplysning/PersonopplysningerServiceTest.java
package no.nav.familie.integrasjoner.personopplysning;

import no.nav.familie.integrasjoner.personopplysning.domene.PersonhistorikkInfo;
import no.nav.familie.integrasjoner.personopplysning.domene.Personinfo;
import no.nav.familie.integrasjoner.personopplysning.domene.TpsOversetter;
import no.nav.familie.integrasjoner.personopplysning.domene.adresse.AdresseType;
import no.nav.familie.integrasjoner.personopplysning.domene.adresse.TpsAdresseOversetter;
import no.nav.familie.integrasjoner.personopplysning.domene.relasjon.Familierelasjon;
import no.nav.familie.integrasjoner.personopplysning.domene.relasjon.RelasjonsRolleType;
import no.nav.familie.integrasjoner.personopplysning.domene.relasjon.SivilstandType;
import no.nav.familie.integrasjoner.personopplysning.domene.status.PersonstatusType;
import no.nav.familie.integrasjoner.personopplysning.domene.tilhørighet.Landkode;
import no.nav.familie.integrasjoner.personopplysning.internal.PersonConsumer;
import no.nav.familie.kontrakter.ks.søknad.testdata.SøknadTestdata;
import no.nav.tjeneste.virksomhet.person.v3.binding.HentPersonPersonIkkeFunnet;
import no.nav.tjeneste.virksomhet.person.v3.binding.HentPersonSikkerhetsbegrensning;
import no.nav.tjeneste.virksomhet.person.v3.binding.HentPersonhistorikkPersonIkkeFunnet;
import no.nav.tjeneste.virksomhet.person.v3.binding.HentPersonhistorikkSikkerhetsbegrensning;
import no.nav.tjeneste.virksomhet.person.v3.feil.PersonIkkeFunnet;
import no.nav.tjeneste.virksomhet.person.v3.feil.Sikkerhetsbegrensning;
import no.nav.tjeneste.virksomhet.person.v3.meldinger.HentPersonRequest;
import no.nav.tjeneste.virksomhet.person.v3.meldinger.HentPersonhistorikkRequest;
import org.assertj.core.api.Assertions;
import org.junit.Before;
import org.junit.Test;
import org.springframework.web.client.HttpClientErrorException;

import java.time.LocalDate;
import java.time.temporal.ChronoUnit;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

public class PersonopplysningerServiceTest {

    private static final String PERSONIDENT = SøknadTestdata.farPersonident;
    private static final LocalDate TOM = LocalDate.now();
    private static final LocalDate FOM = TOM.minusYears(5);

    private PersonConsumer personConsumer;
    private PersonopplysningerService personopplysningerService;

    @Before
    public void setUp() throws Exception {
        personConsumer = new PersonopplysningerTestConfig().personConsumerMock();
        personopplysningerService =
                new PersonopplysningerService(this.personConsumer, new TpsOversetter(new TpsAdresseOversetter()));
    }

    @Test
    public void personhistorikkInfoSkalGiFeilVedUgyldigAktørId() throws Exception {
        when(personConsumer.hentPersonhistorikkResponse(any(HentPersonhistorikkRequest.class)))
                .thenThrow(new HentPersonhistorikkPersonIkkeFunnet("Feil", any(PersonIkkeFunnet.class)));

        assertThatThrownBy(() -> personopplysningerService.hentHistorikkFor(PERSONIDENT, FOM, TOM)).isInstanceOf(
                HttpClientErrorException.NotFound.class);
    }

    @Test
    public void personHistorikkSkalGiFeilVedSikkerhetsbegrensning() throws Exception {
        when(personConsumer.hentPersonhistorikkResponse(any(HentPersonhistorikkRequest.class)))
                .thenThrow(new HentPersonhistorikkSikkerhetsbegrensning("Feil", any(Sikkerhetsbegrensning.class)));

        assertThatThrownBy(() -> personopplysningerService.hentHistorikkFor(PERSONIDENT, FOM, TOM)).isInstanceOf(
                HttpClientErrorException.Forbidden.class);
    }

    @Test
    public void personinfoSkalGiFeilVedUgyldigAktørId() throws Exception {
        when(personConsumer.hentPersonResponse(any(HentPersonRequest.class)))
                .thenThrow(new HentPersonPersonIkkeFunnet("Feil", any(PersonIkkeFunnet.class)));

        assertThatThrownBy(() -> personopplysningerService.hentPersoninfoFor(PERSONIDENT))
                .isInstanceOf(HttpClientErrorException.NotFound.class);
    }

    @Test
    public void personinfoSkalGiFeilVedSikkerhetsbegrensning() throws Exception {
        when(personConsumer.hentPersonResponse(any(HentPersonRequest.class)))
                .thenThrow(new HentPersonSikkerhetsbegrensning("Feil", any(Sikkerhetsbegrensning.class)));

        assertThatThrownBy(() -> personopplysningerService.hentPersoninfoFor(PERSONIDENT))
                .isInstanceOf(HttpClientErrorException.Forbidden.class);
    }

    @Test
    public void skalKonvertereResponsTilPersonInfo() {
        Personinfo response = personopplysningerService.hentPersoninfoFor(PERSONIDENT);

        LocalDate forventetFødselsdato = LocalDate.parse("1990-01-01");

        Familierelasjon barn = response.getFamilierelasjoner().stream()
                                       .filter(p -> p.getRelasjonsrolle() == RelasjonsRolleType.BARN)
                                       .findFirst().orElse(null);
        Familierelasjon ektefelle = response.getFamilierelasjoner().stream()
                                            .filter(p -> p.getRelasjonsrolle() == RelasjonsRolleType.EKTE)
                                            .findFirst().orElse(null);

        assertThat(response.getPersonIdent().getId()).isEqualTo(PERSONIDENT);
        assertThat(response.getStatsborgerskap().erNorge()).isTrue();
        assertThat(response.getSivilstand()).isEqualTo(SivilstandType.GIFT);
        assertThat(response.getAlder()).isEqualTo(ChronoUnit.YEARS.between(forventetFødselsdato, LocalDate.now()));
        Assertions.assertThat(response.getAdresseInfoList()).hasSize(1);
        Assertions.assertThat(response.getPersonstatus()).isEqualTo(PersonstatusType.BOSA);
        assertThat(response.getGeografiskTilknytning()).isEqualTo("0315");
        assertThat(response.getFødselsdato()).isEqualTo(forventetFødselsdato);
        assertThat(response.getDødsdato()).isNull();
        assertThat(response.getDiskresjonskode()).isNull();
        assertThat(response.getAdresseLandkode()).isEqualTo("NOR");

        assertThat(response.getFamilierelasjoner()).hasSize(2);
        assertThat(barn).isNotNull();
        assertThat(barn.getHarSammeBosted()).isTrue();
        assertThat(ektefelle).isNotNull();
        assertThat(ektefelle.getHarSammeBosted()).isTrue();
    }

    @Test
    public void skalKonvertereResponsTilPersonhistorikkInfo() {
        PersonhistorikkInfo response = personopplysningerService.hentHistorikkFor(PERSONIDENT, FOM, TOM);

        assertThat(response.getPersonIdent().getId()).isEqualTo(PERSONIDENT);
        Assertions.assertThat(response.getStatsborgerskaphistorikk()).hasSize(1);
        Assertions.assertThat(response.getPersonstatushistorikk()).hasSize(1);
        Assertions.assertThat(response.getAdressehistorikk()).hasSize(2);

        assertThat(response.getStatsborgerskaphistorikk().get(0).getTilhørendeLand()).isEqualTo(Landkode.NORGE);
        Assertions.assertThat(response.getStatsborgerskaphistorikk().get(0).getPeriode().getFom()).isEqualTo(FOM);

        Assertions.assertThat(response.getPersonstatushistorikk().get(0).getPersonstatus()).isEqualTo(PersonstatusType.BOSA);
        Assertions.assertThat(response.getPersonstatushistorikk().get(0).getPeriode().getTom()).isEqualTo(TOM);

        Assertions.assertThat(response.getAdressehistorikk().get(0).getAdresse().getAdresseType())
                  .isEqualTo(AdresseType.BOSTEDSADRESSE);
        Assertions.assertThat(response.getAdressehistorikk().get(0).getAdresse().getLand()).isEqualTo(Landkode.NORGE.getKode());
        Assertions.assertThat(response.getAdressehistorikk().get(0).getAdresse().getAdresselinje1()).isEqualTo("Sannergata 2");
        Assertions.assertThat(response.getAdressehistorikk().get(0).getAdresse().getPostnummer()).isEqualTo("0560");
        Assertions.assertThat(response.getAdressehistorikk().get(0).getAdresse().getPoststed()).isEqualTo("OSLO");

        Assertions.assertThat(response.getAdressehistorikk().get(1).getAdresse().getAdresseType())
                  .isEqualTo(AdresseType.MIDLERTIDIG_POSTADRESSE_UTLAND);
        Assertions.assertThat(response.getAdressehistorikk().get(1).getAdresse().getLand()).isEqualTo("SWE");
        Assertions.assertThat(response.getAdressehistorikk().get(1).getAdresse().getAdresselinje1()).isEqualTo("TEST 1");
    }

}

File: src/test/java/no/nav/familie/integrasjoner/personopplysning/PersonopplysningerServiceTest.kt
package no.nav.familie.integrasjoner.personopplysning

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import no.nav.familie.integrasjoner.client.rest.PdlClientCredentialRestClient
import no.nav.familie.integrasjoner.client.rest.PdlRestClient
import no.nav.familie.integrasjoner.client.rest.RegoppslagRestClient
import no.nav.familie.integrasjoner.personopplysning.internal.Adressebeskyttelse
import no.nav.familie.integrasjoner.personopplysning.internal.FORELDERBARNRELASJONROLLE
import no.nav.familie.integrasjoner.personopplysning.internal.Fullmakt
import no.nav.familie.integrasjoner.personopplysning.internal.PdlForelderBarnRelasjon
import no.nav.familie.integrasjoner.personopplysning.internal.PdlPersonMedRelasjonerOgAdressebeskyttelse
import no.nav.familie.integrasjoner.personopplysning.internal.Sivilstand
import no.nav.familie.kontrakter.felles.Tema
import no.nav.familie.kontrakter.felles.personopplysning.SIVILSTAND
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class PersonopplysningerServiceTest {

    private lateinit var personopplysningerService: PersonopplysningerService

    private val pdlRestClient = mockk<PdlRestClient>()
    private val pdlClientCredentialRestClient = mockk<PdlClientCredentialRestClient>()
    private val regoppslagRestClient = mockk<RegoppslagRestClient>()

    @BeforeEach
    fun setUp() {
        personopplysningerService = PersonopplysningerService(
            pdlRestClient,
            pdlClientCredentialRestClient,
            regoppslagRestClient,
        )
    }

    @Test
    fun `hentPersonMedRelasjoner skal kalle på pdl 3 ganger, hovedpersonen, relasjonene og barnets andre forelder`() {
        val hovedPerson = "1" to
            lagPdlPersonMedRelasjoner(
                familierelasjoner = listOf(
                    PdlForelderBarnRelasjon(
                        "2",
                        FORELDERBARNRELASJONROLLE.BARN,
                    ),
                ),
                sivilstand = listOf(Sivilstand(SIVILSTAND.GIFT, "3")),
                fullmakt = listOf(Fullmakt("4")),
            )
        val barn = lagPdlPersonMedRelasjoner(
            familierelasjoner = listOf(
                PdlForelderBarnRelasjon(
                    "22",
                    FORELDERBARNRELASJONROLLE.FAR,
                ),
            ),
        )
        every { pdlClientCredentialRestClient.hentPersonMedRelasjonerOgAdressebeskyttelse(any(), any()) } answers {
            firstArg<List<String>>().map { it to if (it == "2") barn else lagPdlPersonMedRelasjoner() }.toMap()
        }
        every { pdlClientCredentialRestClient.hentPersonMedRelasjonerOgAdressebeskyttelse(listOf("1"), any()) } returns mapOf(hovedPerson)

        val hentPersonMedRelasjoner = personopplysningerService.hentPersonMedRelasjoner("1", Tema.ENF)

        assertThat(hentPersonMedRelasjoner.barn.single().personIdent).isEqualTo("2")
        assertThat(hentPersonMedRelasjoner.barnsForeldrer.single().personIdent).isEqualTo("22")
        assertThat(hentPersonMedRelasjoner.sivilstand.single().personIdent).isEqualTo("3")
        assertThat(hentPersonMedRelasjoner.fullmakt.single().personIdent).isEqualTo("4")

        verify(exactly = 1) {
            pdlClientCredentialRestClient.hentPersonMedRelasjonerOgAdressebeskyttelse(listOf("1"), any())
            pdlClientCredentialRestClient.hentPersonMedRelasjonerOgAdressebeskyttelse(listOf("2", "3", "4"), any())
            pdlClientCredentialRestClient.hentPersonMedRelasjonerOgAdressebeskyttelse(listOf("22"), any())
        }
    }

    private fun lagPdlPersonMedRelasjoner(
        familierelasjoner: List<PdlForelderBarnRelasjon> = emptyList(),
        sivilstand: List<Sivilstand> = emptyList(),
        fullmakt: List<Fullmakt> = emptyList(),
        adressebeskyttelse: List<Adressebeskyttelse> = emptyList(),
    ) =
        PdlPersonMedRelasjonerOgAdressebeskyttelse(familierelasjoner, sivilstand, fullmakt, adressebeskyttelse)
}
